tests/unit/test_main.py

```diff
@@ -99,46 +103,126 @@ def test_log_or_exit_on_exceptions_debug(plugin_exc_mock, mocker, monkeypatch):
 
     errors = [('bad.plugin', plugin_exc_mock)]
 
-    main._log_or_exit_on_exceptions(errors, debug=True)
+    main._log_or_exit_on_exceptions('base msg', errors, debug=True)
 
     logging_mock.warn.assert_called_once()
     logging_mock.error.assert_not_called()
 
 
-run_args = 'has_active_plugins,exp_log_count'
-run_params = [
-    (True, 2),
-    (False, 1),
-]
+@pytest.fixture
+def mock_log_or_exit_on_exc(mocker, monkeypatch):
+    mock = mocker.Mock()
+    monkeypatch.setattr('gordon.main._log_or_exit_on_exceptions', mock)
+    return mock
+
+
+@pytest.mark.parametrize('plugins,exp_log_calls', (
+    (('event_consumer', 'enricher', 'publisher', 'runnable'), 0),
+    (('event_consumer', 'enricher', 'publisher'), 0),
+    (('event_consumer', 'enricher'), 1),
+    (('event_consumer',), 1),
+    (('runnable',), 1),
+
+))
+def test_assert_required_plugins(plugins, exp_log_calls, inited_plugins,
+                                 mock_log_or_exit_on_exc):
+    """Assert required plugins are installed, else warn/error out."""
+    plugins = [inited_plugins[p] for p in plugins]
+    main._assert_required_plugins(plugins, debug=True)
+
+    assert exp_log_calls == mock_log_or_exit_on_exc.call_count
+
+
+@pytest.mark.parametrize('has_run,has_run_async,exp_plugins,exp_mock_call', (
+    (True, True, 2, 0),  # has `async def run` method
+    (True, False, 1, 1),  # has `def run` method - not async
+    (False, False, 1, 1),  # does not have a any `run` method
+))
+def test_gather_runnable_plugins(has_run, has_run_async, exp_plugins,
+                                 exp_mock_call, monkeypatch,
+                                 inited_plugins, mock_log_or_exit_on_exc):
+    """Return all plugins with an async `run` method."""
+    patch_run = 'tests.unit.conftest.EventConsumerStub.run'
```

Comment:
This is the bit I meant about async.  We should also test that it does nothing if a runnable's run() isn't async.  Or isn't present, for that matter.

Reply:
Also, if the runnable has a run() but it's not async, should we surface that as a warning or something?

Reply:
Maybe I'm not understanding correctly, but isn't this case tested in the [second iteration](https://github.com/spotify/gordon/pull/17/commits/2733fc4e7bcdb198154344fd2c23b8418b4d6c61#diff-a324e09b40482c4a38ed3c5fd93c0f4eR138) of this test? There's only one runnable plugin when four are available, two having a run method, and only one of those two actually async.

A warning on a sync run is a good suggestion.

Reply:
Actually, re: warning - we only get into this case because we don't yet verify interface implementations, which is a v2 thing.

Reply:
We're only patching one of the interfaced plugins.  I was suggesting having tests on the non-interface "runnable" ones that we identify based just on their having an async run().